src/express/common/milestone-utils.js

```diff
@@ -0,0 +1,396 @@
+      for (let i = 1; i < validators.length; i++) {
+        if (
+          String(proposer).toLowerCase() ===
+          String(validators[i].address).toLowerCase()
+        ) {
+          console.log('📍Validated milestone proposer')
+          return
+        }
+      }
+
+      console.log(
+        '📍Invalid milestone got proposed from validator/s of cluster 1'
+      )
+      console.log(
+        '📍Milestone proposer:',
+        proposer,
+        ', validators: ',
+        validators
+      )
+    }
+  } catch (error) {
+    console.log(
+      '📍Error in validating milestone proposer, skipping check. Error:',
+      error
+    )
+  }
+}
+
+export async function removePeers(ip, peers) {
+  const tasks = []
+  for (let i = 0; i < peers.length; i++) {
+    const command = `~/go/bin/bor attach ~/.bor/data/bor.ipc --exec "admin.removePeer('${peers[i]}')"`
+    tasks.push(runSshCommand(ip, command, maxRetries))
+  }
+
+  let response = false
+  await Promise.all(tasks)
+    .then(() => {
+      response = true
+    })
+    .catch((error) => {
+      console.log('📍Unable to remove peers, error:', error)
+    })
+
+  return response
+}
+
+export async function addPeers(ip, peers) {
+  const tasks = []
+  for (let i = 0; i < peers.length; i++) {
+    const command = `~/go/bin/bor attach ~/.bor/data/bor.ipc --exec "admin.addPeer('${peers[i]}')"`
+    tasks.push(runSshCommand(ip, command, maxRetries))
+  }
+
+  let response = false
+  await Promise.all(tasks)
+    .then(() => {
+      response = true
+    })
+    .catch((error) => {
+      console.log('📍Unable to add peers, error:', error)
+    })
+
+  return response
+}
+
+export async function getPeerLength(ip) {
+  const command =
+    '/home/ubuntu/go/bin/bor attach ~/.bor/data/bor.ipc --exec "admin.peers.length"'
+  try {
+    const length = await runSshCommandWithReturn(ip, command, maxRetries)
+    return parseInt(length)
+  } catch (error) {
+    console.log('📍Unable to query peer length, error:', error)
+  }
+
+  return -1
+}
+
+export async function joinAllPeers(ips, enodes) {
+  let tasks = []
+  for (let i = 0; i < ips.length; i++) {
+    tasks.push(addPeers(ips[i], enodes))
+  }
+
+  await Promise.all(tasks).then((values) => {
+    if (values.includes(false) || values.includes(undefined)) {
+      console.log('📍Failed to add peers for rejoining clusters')
+      return false
+    }
+  })
+
+  // Validate by fetching peer length
+  tasks = []
+  for (let i = 0; i < ips.length; i++) {
+    tasks.push(getPeerLength(ips[i]))
+  }
+
+  let lengths
+  await Promise.all(tasks).then((values) => {
+    lengths = values
+  })
+
+  if (!lengths) {
+    await timer(100)
+    if (!lengths) {
+      return false
+    }
+  }
+
+  if (lengths.includes(-1)) {
+    console.log('📍Unable to query peer length, exiting')
+    return false
+  }
+
+  console.log('📍Peer length:', lengths)
+  return true
+}
+
+export async function createClusters(ips, enodes, split = 1) {
```

Comment:
Is this function the only one needed to create a deterministic fork on the devnet?
I know it's out of scope for your PR, but was wondering if I can creare a command for `express-cli` out of it, something like `--reorg [split]`

Reply:
Oh interesting. Yeah we could actually do so.

Reply:
Nice, thanks!
If you want, once other comments are addressed, I can work on it and then submit to you for review